docs/source/chunk.rst

```diff
@@ -0,0 +1,352 @@
+          # We need to know the exact amount of memory to preallocate it
+          # in shared memory so the main process can access it quickly
+          # Send each tensor individually to shared memory leads to 2x perf hit
+          features = []
+          features_numel = 0
+          features_out = None
+          labels = []
+          labels_numel = 0
+          labels_out = None
+
+          for b in batch:
+              # C++ FooExampleType to Pytorch Tensors (zero-copy)
+              feature_tensor = torch.from_numpy(numpy.array(b.feature_))
+              label_tensor = torch.from_numpy(numpy.array(b.label_))
+
+              # Determining tensor size for using shared memory
+              features_numel += feature_tensor.numel()
+              features.append(feature_tensor)
+              labels_numel += label_tensor.numel()
+              labels.append(label_tensor)
+
+          # Allocating shared memory
+          features_shared_storage = features[0].storage(
+          )._new_shared(features_numel)
+          features_out = features[0].new(features_shared_storage)
+          labels_shared_storage = labels[0].storage()._new_shared(labels_numel)
+          labels_out = labels[0].new(labels_shared_storage)
+
+          # Stacking (copying) tensors into shared memory
+          torch.stack(features, out=features_out)
+          torch.stack(labels, out=labels_out)
+
+          # Return shared memory on torch.multprocessing queues (zero-copy)
+          dict['feature'] = features_out
+          dict['label'] = labels_out
+          return dict
+
+  def worker_init_fn(worker_id):
+    # A recent change on pytorch enabled multithreading by default
+    # Dataloader logic requires a single thread, though
+    # Until the https://github.com/pytorch/pytorch/issues/19213 is resolved,
+    # you have to create an environment variable OMP_NUM_THREADS=1 as a workaround
+    torch.set_num_threads(1)
+    dataset = torch.utils.data.get_worker_info().dataset
+    dataset.set_chunk_sampler_stride(worker_id)
+    dataset.reset()
+
+  batch_size = 5
+  cache_size = 100
+  preloaders = 1
+  num_workers = 2
+  reader = cpp_extension.FooChunkDataReader()
+  opt = chunk.ChunkDatasetOptions(preloader_count=preloaders, batch_size=batch_size, cache_size=cache_size)
+
+  foo_chunkdataset = cpp_extension.FooChunkDataset(chunk_reader=reader,
+                                                   options=opt,
+                                                   shuffle_chunks=False,
+                                                   shuffle_samples=False)
+
+  trainset = chunk.ChunkDatasetWrapper(foo_chunkdataset)
+  trainset.reset()
+  trainloader = DataLoader(dataset=trainset,
+                           num_workers=num_workers,
+                           collate_fn=collate_fn,
+                           worker_init_fn=worker_init_fn)
+  for i, batch in enumerate(trainloader, 0):
+    print('Batch {}: {}'.format(i, batch))
```

Comment:
I really think that given the ratio of code to text in here, this should really become an example repository and not part of our docs...